p2p/README.md

```diff
@@ -67,105 +66,431 @@ flowchart TD
 
-### Raintree Router Architecture
+## Module Architecture
 
-_DISCUSS(team): If you feel this needs a diagram, please reach out to the team for additional details._
-_TODO(olshansky, BenVan): Link to RainTree visualizations once it is complete._
+### Legends
+```mermaid
+flowchart
+subgraph Legend
+    m[[`Method`]]
+    c[Component]
+
+    m -- "unconditional usage" --> c
+    m -. "conditional usage" .-> c
+    m -. "ignored" .-x c
+end
+```
+
+```mermaid
+classDiagram
+class ConcreteType {
+  +ExportedField
+  -unexportedField
+  +ExportedMethod(argType) returnType
+  -unexportedMethod()
+}
+
+class InterfaceType {
+    <<interface>>
+    +Method(argType) (returnType1, returnType2)
+}
+
+ConcreteType --|> InterfaceType : Interface realization
+
+ConcreteType --> OtherType : Direct usage
+ConcreteType --o OtherType : Composition
+ConcreteType --* OtherType : Aggregatation
+ConcreteType ..*  "(cardinality)" OtherType : Indirect (via interface)
+```
+
+### P2P Module / Router Decoupling
 
-### Message Propagation
+The P2P module encapsulates the `RaiTreeRouter` and `BackgroundRouter` submodules.
+The P2P module internally refers to these as the `stakedActorRouter` and `unstakedActorRouter`, respectively.
 
-Given `Local P2P Module` has a message that it needs to propagate:
+Depending on the necessary routing scheme (unicast / broadcast) and whether the peers involved are staked actors, a node will use one or both of these routers.
 
-<ul style="list-style-type: none;">
-    <li>1a. <code>Raintree Router</code> selects targets from the <code>Pokt Peerstore</code>, <strong>which only includes staked actors</strong></li>
-    <li>1b. <code>Background Router</code> selects targets from the libp2p <code>Peerstore</code>, <strong>which includes all P2P participants</strong></li>
-    <li>2. Libp2p <code>Host</code> manages opening and closing streams to targeted peers</li>
-    <li>3. <code>Remote P2P module</code>'s (i.e. receiver's) <code>handleStream</code> is called (having been registered via <code>setStreamHandler()</code>)</li>
-    <li>4a. <code>handleStream</code> propagates message via <code>Raintree Router</code></li>
-    <li>4b. <code>handleStream</code> propagates message via <code>Background Router</code></li>
-    <li>5a. Repeat step 1a from <code>Remote P2P Module</code>'s perspective targeting its next peers</li>
-    <li>5b. Repeat step 1b from <code>Remote P2P Module</code>'s perspective targeting its next peers</li>
-</ul>
+**Unicast**
+
+| Sender         | Receiver       | Router          | Example Usage                                        |
+|----------------|----------------|-----------------|------------------------------------------------------|
+| Staked Actor   | Staked Actor   | Raintree only   | Consensus (state sync) messages (to validators only) |
```

Comment:
@Olshansk would you mind elaborating on this suggestion?

Reply:
As noted in the P2P knowledge transfer (#438) ticket, for hotstuff consensus, we don't use RainTree for messages propagation between staked actors (e.g. state sync) but for consensus, it's a star pattern (i.e. direct)

Reply:
Gotcha, thanks for the additional context - I see what you mean now. I think your suggestion of "NA (Direct Send)" is orthogonal to the "router" degree of freedom as it would apply to the entire table. It seems like the "unicast" label above the table was insufficient in the context I intended it to provide. :sweat_smile:

I think where tripping over the distinction between "raintree router" and the "raintree algorithm/propogation". What I was originally trying to communicate here is that the "raintree router" (hence, the column name) will be used for staked-to-staked actor communication. This also aligns with the structure of the component diagrams under "Message Propagation & Handling". 

Do you have a suggestion for how to better structure this?

Reply:
How about this:

```
 Sender         | Receiver       | Router          | Example Usage                                        |
|----------------|----------------|-----------------|------------------------------------------------------|
| Staked Actor   | Staked Actor   | No router  | Consensus (hotstuff) messages (to validators only) |
| Staked Actor   | Staked Actor   | Raintree only   | Consensus (state sync) messages (to validators only) |
| Unstaked Actor | Staked Actor   | Background only | Consensus (state sync) messages (to validators only) |
| Unstaked Actor | Unstaked Actor | Background only | Consensus (state sync) & Debug (CLI) messages        |
```

Reply:
:thinking: I think we're still coming at this from different perspectives, I think I was also confused about the "validator only" distinction.

It seems like you're trying to articulate a couple of points: 

1. That hotstuff messages are unicast rather than broadcast
2. Unicast routing does not use the raintree gossip algorithm

In my view, 1. is implied by the fact that this is the "unicast" table and therefore applies to all these. If we applied your perspective consistently to the entire table, I would expect to see "no router" in the "router" column for every row, which seems wrong to me.

My intention with this table was not to speak to which "gossip algorithm" is used, but rather which `Router` implementation is handling the control flow, matching the diagrams that follow. This is important because the raintree and background routers have distinct message types, protocol IDs, and stream handlers.

I hope that helps clarify my perspective. Please correct my understanding of yours if I've missed something or misunderstood.

I've updated both tables to try to improve clarity as well as fix some mistakes / fill in gaps, here's my new proposal:

**Unicast**

| Sender         | Receiver       | Router          | Example Usage                                                        |
|----------------|----------------|-----------------|----------------------------------------------------------------------|
| Staked Actor   | Staked Actor   | Raintree only   | Consensus hotstuff messages (validators only) & state sync responses |
| Staked Actor   | Untaked Actor  | Background only | Consensus state sync responses                                       |
| Unstaked Actor | Staked Actor   | Background only | Consensus state sync responses, debug messages                       |
| Unstaked Actor | Unstaked Actor | Background only | Consensus state sync responses, debug messages                       |

**Broadcast**

| Broadcaster    | Receiver       | Router                | Example Usage                                                   |
|----------------|----------------|-----------------------|-----------------------------------------------------------------|
| Staked Actor   | Staked Actor   | Raintree + Background | Utility tx messages, consensus state sync requests              |
| Staked Actor   | Untaked Actor  | Background only       | Utility tx messages (redundancy), consensus state sync requests |
| Unstaked Actor | Staked Actor   | Background only       | Utility tx messages (redundancy), consensus state sync requests |
| Unstaked Actor | Unstaked Actor | Background only       | Utility tx messages, consensus state sync requests              |